ffmpeg/lpms_ffmpeg.c

```diff
@@ -1173,39 +1176,64 @@ int process_out(struct input_ctx *ictx, struct output_ctx *octx, AVCodecContext
     return encode(encoder, inf, octx, ost);
   }
 
+  int is_video = (AVMEDIA_TYPE_VIDEO == ost->codecpar->codec_type);
   // Sometimes we have to reset the filter if the HW context is updated
   // because we initially set the filter before the decoder is fully ready
   // and the decoder may change HW params
-  if (AVMEDIA_TYPE_VIDEO == ost->codecpar->codec_type &&
-      inf && inf->hw_frames_ctx && filter->hwframes &&
+  if (is_video && inf && inf->hw_frames_ctx && filter->hwframes &&
       inf->hw_frames_ctx->data != filter->hwframes) {
     free_filter(&octx->vf); // XXX really should flush filter first
     ret = init_video_filters(ictx, octx);
     if (ret < 0) return lpms_ERR_FILTERS;
   }
-  // Start filter flushing process if necessary
-  if (!inf && !filter->flushed) {
-    // Set input frame to the last frame
-    int is_video = AVMEDIA_TYPE_VIDEO == ost->codecpar->codec_type;
-    AVFrame *frame = is_video ? ictx->last_frame_v : ictx->last_frame_a;
-    // Increment pts offset by:
-    // - input packet's duration usually
-    int64_t dur = frame->pkt_duration;
-    // - output packet's duration if using fps filter & haven't encoded anything yet
-    if (is_video && octx->fps.den && !octx->res->frames) {
-      AVStream *ist = ictx->ic->streams[ictx->vi];
-      dur = av_rescale_q(frame->pkt_duration, ist->r_frame_rate, octx->fps);
+
+  if (is_video) {
+    // Video Stream
+    AVStream *ist = ictx->ic->streams[ictx->vi];
+    int input_dur = av_rescale_q(1, av_inv_q(ist->r_frame_rate), ist->time_base);
+    if (inf) {
+      // Set custom PTS, Set opaque as this is non-flush packet
+      inf->opaque = (void *) inf->pts;
+      if (octx->fps.den)
+        filter->custom_pts += input_dur;
+      else
+        filter->custom_pts = inf->pts;
+    } else if (!filter->flushed) {
+      // Video Flushing
+      AVFrame *frame = ictx->last_frame_v;
+      if (octx->fps.den && !octx->res->frames) {
+        // Haven't encoded anything yet, so need to force flush by rescaling PTS to match output
+        input_dur = av_rescale_q(input_dur, ist->r_frame_rate, octx->fps);
+      } else if (!octx->fps.den) {
+        // Passthrough, use packet duration instead of custom duration
+        input_dur = frame->pkt_duration;
+      }
+      filter->custom_pts += input_dur;
+      inf = frame;
+      inf->opaque = NULL;
+      is_flushing = 1;
+    }
+  } else {
+    // Audio Stream
+    if (inf) {
+      inf->opaque = (void *) inf->pts;
+      filter->custom_pts = inf->pts;
+    } else if (!filter->flushed) {
+      // Audio Flushing
+      AVFrame *frame = ictx->last_frame_a;
+      filter->custom_pts += frame->pkt_duration;
+      inf = frame;
+      inf->opaque = NULL;
+      is_flushing = 1;
```

Comment:
Ah now that I read this, it can be simplified a lot. Will do it in next iteration.